homeassistant/components/obihai/sensor.py

```diff
@@ -35,46 +32,90 @@
+        is_fixable=False,
+        severity=issue_registry.IssueSeverity.ERROR,
+        translation_key="manual_migration",
+    )
+
+    if discovery_info:
+        config = PLATFORM_SCHEMA(discovery_info)
+
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN,
+            context={"source": SOURCE_IMPORT},
+            data=config,
+        )
+    )
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up the Obihai sensor entries."""
 
-    username = config[CONF_USERNAME]
-    password = config[CONF_PASSWORD]
-    host = config[CONF_HOST]
+    username = entry.data[CONF_USERNAME]
+    password = entry.data[CONF_PASSWORD]
+    host = entry.data[CONF_HOST]
+    requester = ObihaiConnection(host, username, password)
 
-    sensors = []
+    await hass.async_add_executor_job(requester.update)
+    sensors = requester.sensors
 
-    pyobihai = PyObihai(host, username, password)
+    async_add_entities(sensors, update_before_add=True)
 
-    login = pyobihai.check_account()
-    if not login:
-        _LOGGER.error("Invalid credentials")
-        return
 
-    serial = pyobihai.get_device_serial()
+class ObihaiConnection:
+    """Contains a list of Obihai Sensors."""
 
-    services = pyobihai.get_state()
+    def __init__(
+        self,
+        host: str,
+        username: str = DEFAULT_USERNAME,
+        password: str = DEFAULT_PASSWORD,
+    ) -> None:
+        """Store configuration."""
+        self.sensors: list[ObihaiServiceSensors] = []
+        self.host = host
+        self.username = username
+        self.password = password
 
-    line_services = pyobihai.get_line_state()
+    def update(self) -> bool:
+        """Validate connection and retrieve a list of sensors."""
+        pyobihai = get_pyobihai(self.host, self.username, self.password)
 
-    call_direction = pyobihai.get_call_direction()
+        if not pyobihai.check_account():
+            return False
 
-    for key in services:
-        sensors.append(ObihaiServiceSensors(pyobihai, serial, key))
+        serial = pyobihai.get_device_serial()
+        services = pyobihai.get_state()
+        line_services = pyobihai.get_line_state()
+        call_direction = pyobihai.get_call_direction()
 
-    if line_services is not None:
-        for key in line_services:
-            sensors.append(ObihaiServiceSensors(pyobihai, serial, key))
+        for key in services:
+            self.sensors.append(ObihaiServiceSensors(pyobihai, serial, key))
 
-    for key in call_direction:
-        sensors.append(ObihaiServiceSensors(pyobihai, serial, key))
+        if line_services is not None:
+            for key in line_services:
+                self.sensors.append(ObihaiServiceSensors(pyobihai, serial, key))
 
-    add_entities(sensors)
+        for key in call_direction:
+            self.sensors.append(ObihaiServiceSensors(pyobihai, serial, key))
```

Comment:
I think these 7 lines should be moved to `async_setup_entry`, replacing `sensors = requester.sensors`
Then you can move `ObihaiConnection` to the api module, and expose `serial`/`services`/`line_services`/`call_direction` as properties